Add `Dataset.drop_dims`

kmsquire opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

 - [x] Closes #1949 
 - [x] Tests added
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

pep8speaks commented:
Hello @kmsquire! Thanks for submitting the PR.

 - In the file [`xarray/core/dataset.py`](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/core/dataset.py), following are the PEP8 issues :

> [Line 2833:80](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/core/dataset.py#L2833): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

 - In the file [`xarray/tests/test_dataset.py`](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py), following are the PEP8 issues :

> [Line 1872:34](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1872): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1872:38](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1872): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1879:37](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1879): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1880:34](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1880): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1880:38](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1880): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1890:34](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1890): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1890:38](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1890): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1894:30](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1894): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1895:34](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1895): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1895:38](https://github.com/pydata/xarray/blob/14853f00a9be28ea54f3b8a77fbc5986fe67a5cb/xarray/tests/test_dataset.py#L1895): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

kmsquire commented:
I was looking for this functionality and ran across #1949.

There was one bit of behavior that was unclear to me.  Say I have

```python
In [6]: data = xr.Dataset({'A': (['x', 'y'], np.random.randn(2, 3)),
   ...:                    'B': ('x', np.random.randn(2)),
   ...:                    'x': ['a', 'b'], 'z': np.pi})

In [7]: data
Out[7]:
<xarray.Dataset>
Dimensions:  (x: 2, y: 3)
Coordinates:
  * x        (x) <U1 'a' 'b'
Dimensions without coordinates: y
Data variables:
    A        (x, y) float64 -0.662 -0.705 -0.1866 -0.8655 0.5816 -0.2308
    B        (x) float64 -0.1002 2.402
    z        float64 3.142

In [8]: data.drop_dims('x')
Out[8]:
<xarray.Dataset>
Dimensions:  ()
Data variables:
    z        float64 3.142
```

In this case, I dropped `x`, but because that eliminated variable `A`, I also dropped my `y` dimension, since `A` was the only variable using `y`.

At one point, I implemented this slightly differently, which instead gave

```python
In [8]: data.drop_dims('x')
Out[8]:
<xarray.Dataset>
Dimensions:  (y: 3)
Dimensions without coordinates: y
Data variables:
    z        float64 3.142
```

In some ways, this seems preferable, especially if `y` has coordinates.  The only downside I can see is that this version called `Dataset._replace_vars_and_dims()`, the docstring for which says it's deprecated (although it's still widely used).

kmsquire commented:
@shoyer I'm sure you're busy... wondering if you might have time to review this (or direct it to someone else to review)? 
Other than my question above, it should be pretty straightforward.

shoyer commented:
It's OK to use `_replace_with_new_dims` still. I think we'll eventually replace it in favor of `_replace_with_new_dims` but we aren't there yet.

I agree that it makes sense to recompute dimensions from variables. There's not much point in keeping around extra dimensions that don't have corresponding variables, and I'm pretty sure that this is consistent with how `Dataset.drop()` works.

kmsquire commented:
> I agree that it makes sense to recompute dimensions from variables.

Okay, that's what this PR already did (via the call to `_replace_with_new_dims`).

Rebased on master.  From my perspective, this should be good to go.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -2797,6 +2797,42 @@ def _drop_vars(self, names):
         coord_names = set(k for k in self._coord_names if k in variables)
         return self._replace_vars_and_dims(variables, coord_names)
 
+    def drop_dims(self, drop_dims, inplace=False):
```
We've deprecated the `inplace` option for other Dataset methods, so I would rather not add it here.

kmsquire replied:
Okay.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -2797,6 +2797,42 @@ def _drop_vars(self, names):
         return self._replace_vars_and_dims(variables, coord_names)
 
+    def drop_dims(self, drop_dims, inplace=False):
+        """Drop dimensions from a Dataset.  Any data variables containing
```
For something that fits on one line, maybe:
`"""Drop dimensions and associated variables from this dataset."""`

kmsquire replied:
Okay.

shoyer commented on doc/indexing.rst:
```diff
@@ -229,7 +229,7 @@ arrays). However, you can do normal indexing with dimension names:
 
-Dropping labels
+Dropping labels and dimensions
 ---------------
```
I think you need to increase the width of the `----` line to match the new label.

kmsquire replied:
Done.

shoyer commented on doc/indexing.rst:
```diff
@@ -241,6 +241,15 @@ index labels along a dimension dropped:
+
+    ds.drop_dims('time')
+
+``drop_dims`` is a ``Dataset`` method.
```
Maybe delete this, and above say "to drop a full dimension from a dataset."?

kmsquire replied:
Done.

kmsquire commented:
@shoyer I force pushed the requested changes.  Please let me know if there is anything else.  Thanks!

shoyer commented on xarray/core/dataset.py:
```diff
@@ -2797,6 +2797,38 @@ def _drop_vars(self, names):
+        coord_names = set(k for k in self._coord_names if k in variables)
+
+        return self._replace_with_new_dims(
+            variables, coord_names, inplace=inplace)
```
`inplace` is now an undefined variable, so your tests are failing

kmsquire replied:
Whoops!  I took that out of the tests, but forgot to take it out of `dataset.py`.  Should be fixed now.

kmsquire commented:
Rebased to fix the NEWS conflict.

shoyer commented:
thanks @kmsquire, especially for your patience here!

kmsquire commented:
No worries--thanks for merging!

naught101 commented:
Is there currently a way to drop unused coordinates, like this?

naught101 commented:
I guess `ds = ds.drop([c for c in ds.coords if c not in ds.dims])` works. Might be nice to have a convenience function for that though.
